src/main/java/org/sonatype/repository/chef/internal/util/ChefDataAccess.java

```diff
@@ -0,0 +1,149 @@
+import org.sonatype.nexus.repository.view.payloads.StreamPayload;
+import org.sonatype.nexus.repository.view.payloads.StreamPayload.InputStreamSupplier;
+
+import com.google.common.base.Supplier;
+import com.google.common.collect.ImmutableList;
+
+import static java.util.Collections.singletonList;
+import static org.sonatype.nexus.repository.storage.ComponentEntityAdapter.P_VERSION;
+import static org.sonatype.nexus.repository.storage.MetadataNodeEntityAdapter.P_NAME;
+
+/**
+ * @since 0.0.1
+ */
+@Named
+public class ChefDataAccess
+{
+  public static final List<HashAlgorithm> HASH_ALGORITHMS = ImmutableList.of(HashAlgorithm.SHA1);
+
+  /**
+   * Find a component by its name and tag (version)
+   *
+   * @return found component or null if not found
+   */
+  @Nullable
+  public Component findComponent(final StorageTx tx,
+                                 final Repository repository,
+                                 final String name,
+                                 final String version)
+  {
+    Iterable<Component> components = tx.findComponents(
+        Query.builder()
+            .where(P_NAME).eq(name)
+            .and(P_VERSION).eq(version)
+            .build(),
+        singletonList(repository)
+    );
+    if (components.iterator().hasNext()) {
+      return components.iterator().next();
+    }
+    return null;
+  }
+
+  /**
+   * Find an asset by its name.
+   *
+   * @return found asset or null if not found
+   */
+  @Nullable
+  public Asset findAsset(final StorageTx tx, final Bucket bucket, final String assetName) {
+    return tx.findAssetWithProperty(MetadataNodeEntityAdapter.P_NAME, assetName, bucket);
+  }
+
+  /**
+   * Save an asset and create blob.
+   *
+   * @return blob content
+   */
+  public Content saveAsset(final StorageTx tx,
+                           final Asset asset,
+                           final Supplier<InputStream> contentSupplier,
+                           final Payload payload) throws IOException
+  {
+    AttributesMap contentAttributes = null;
+    String contentType = null;
+    if (payload instanceof Content) {
+      contentAttributes = ((Content) payload).getAttributes();
+      contentType = payload.getContentType();
+    }
+    return saveAsset(tx, asset, contentSupplier, contentType, contentAttributes);
+  }
+
+  /**
+   * Save an asset and create blob.
+   *
+   * @return blob content
+   */
+  public Content saveAsset(final StorageTx tx,
+                           final Asset asset,
+                           final Supplier<InputStream> contentSupplier,
+                           final String contentType,
+                           @Nullable final AttributesMap contentAttributes) throws IOException
+  {
+    Content.applyToAsset(asset, Content.maintainLastModified(asset, contentAttributes));
+    AssetBlob assetBlob = tx.setBlob(
+        asset, asset.name(), contentSupplier, HASH_ALGORITHMS, null, contentType, false
+    );
+    asset.markAsDownloaded();
```

Comment:
Note: There are changes coming to how this is handled, it will still work but just letting you know for the future - https://github.com/sonatype/nexus-internal/pull/3666